Validations and model specs

mpa-mxiang opened the pull request:
For this PR, I:

- Add validations for
    - user
    - post

mpa-mxiang committed:
Add Validations to the Models

mpa-mxiang committed:
Generate model specs for User and Post

mpa-mxiang committed:
build gem for tests and create tests files

mpa-mxiang committed:
generate RSpec configuration

mpa-mxiang committed:
fix tests and pass

mpa-mxiang committed:
autofix rubocop errors

Reem-lab commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,23 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  it 'is valid with a name and a non-negative integer posts_counter' do
+    user = User.new(name: 'John', posts_counter: 0)
+    expect(user).to be_valid
+  end
+
+  it 'is invalid without a name' do
+    user = User.new(posts_counter: 0)
+    expect(user).to_not be_valid
+  end
+
+  it 'is invalid with a negative posts_counter' do
+    user = User.new(name: 'John', posts_counter: -1)
+    expect(user).to_not be_valid
+  end
+
+  it 'is valid with a nil posts_counter' do
+    user = User.new(name: 'John', posts_counter: nil)
```
- [x] Great job working on the tests so far but I think it would be great if you add tests for all the methods in the User model as mentioned in the requirements since it is very important to test if all the methods in our app are working properly ✅💪🏻🔥🤩 since I just could see tests for the validation so kindly add the rest of tests 👍🏻

- You should add tests for these methods `recent_posts` 🔥


![image](https://github.com/mpa-mxiang/Blog-app/assets/58553711/7e65883e-bf56-4da8-87d5-a5f9e3b0be88)

Reem-lab commented on spec/models/post_spec.rb:
```diff
@@ -0,0 +1,78 @@
+      title: title,
+      comments_counter: 0,
+      likes_counter: 0,
+      author: user
+    )
+    expect(post).to_not be_valid
+  end
+
+  it 'is invalid with a negative comments_counter' do
+    user = User.create(name: 'John', posts_counter: 0)
+    post = Post.new(
+      title: 'Title',
+      comments_counter: -1,
+      likes_counter: 0,
+      author: user
+    )
+    expect(post).to_not be_valid
+  end
+
+  it 'is invalid with a negative likes_counter' do
+    user = User.create(name: 'John', posts_counter: 0)
+    post = Post.new(
+      title: 'Title',
+      comments_counter: 0,
+      likes_counter: -1,
+      author: user
+    )
+    expect(post).to_not be_valid
+  end
+
+  it 'is valid with nil comments_counter and likes_counter' do
+    user = User.create(name: 'John', posts_counter: 0)
+    post = Post.new(
+      title: 'Title',
+      comments_counter: nil,
+      likes_counter: nil,
+      author: user
+    )
+    expect(post).to be_valid
+  end
```
- [x] Great job working on the tests so far but I think it would be great if you add tests for all the methods in the Post model as mentioned in the requirements since it is very important to test if all the methods in our app are working properly ✅💪🏻🔥🤩 since I just could see tests for the validation so kindly add the rest of tests 👍🏻

- You should add tests for these methods `increment_user_posts_count` and `recent_comments` 🔥

![image](https://github.com/mpa-mxiang/Blog-app/assets/58553711/7e65883e-bf56-4da8-87d5-a5f9e3b0be88)

Reem-lab commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,23 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  it 'is valid with a name and a non-negative integer posts_counter' do
+    user = User.new(name: 'John', posts_counter: 0)
```
- [ ] Great job for adding the tests so far for the User and Post models but I think you forgot to test comments and like the model's methods so kindly add tests for them so please add the test for increment_post_comments_count in the comment model 🔥👍🏻 and `update_likes_count` in Like model. as mentioned in the requirements so we should test all the methods models. 👍🏻

Reem-lab reviewed (changes requested):
# Hi @mpa-mxiang, 👋🏻👋🏻

Great job so far! ✨⭐🤩✅
There are some issues that you still need to work on to go to the next project but you are almost there!
<img src="https://media1.giphy.com/media/t8YPf3ACKeouLzEIvr/giphy.gif"/>


## Highlights
- [x] No linters error ✅
- [x] Followed Gitflow ⭐
- [ ] Amazing job working on the functionality for the validations 🔥
- [ ] All tests are passed ✅

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

mpa-mxiang committed:
add test for recent_posts method

mpa-mxiang committed:
add test for increment_user_posts_count method

mpa-mxiang committed:
add test for recent_comment method

mpa-mxiang committed:
create comment test file

mpa-mxiang committed:
fix comment tests

delete comment text empty case since there is no related validation in comment.rb, add two invalidation cases and two method test cases

mpa-mxiang committed:
add tests for like class

mpa-mxiang committed:
autofix rubocop errors

Danieldotcomcoder reviewed (approved):
Hello @mpa-mxiang , 

Your project is complete!  There is nothing else to say other than... it's time to merge it 👍 👍 😄 👏 

Congratulations! 🎉 🎉 🎉 

## Highlights
- Clear and nice code ✔✔
- Tests are good and passing ✔✔ 

![comp-2](https://user-images.githubusercontent.com/87448628/159119684-e86473c4-9b60-43a3-90f2-54048e841394.gif)
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

mpa-mxiang commented:
Thank you @Danieldotcomcoder  for the approval!
